src/btree/bt_sync.c

```diff
@@ -22,30 +22,36 @@ __sync_checkpoint_can_skip(WT_SESSION_IMPL *session, WT_REF *ref)
 
     WT_ASSERT_SPINLOCK_OWNED(session, &S2BT(session)->flush_lock);
 
-    mod = ref->page->modify;
-    txn = session->txn;
-
     /*
-     * We can skip some dirty pages during a checkpoint. The requirements:
+     * Don't skip checkpoint if at least one of the following conditions is met:
      *
-     * 1. Not a history btree. As part of the checkpointing the data store, we will move the older
-     *    values into the history store without using any transactions. This led to representation
-     *    of all the modifications on the history store page with a transaction that is maximum than
-     *    the checkpoint snapshot. But these modifications are done by the checkpoint itself, so we
-     *    shouldn't ignore them for consistency.
-     * 2. they must be leaf pages,
-     * 3. there is a snapshot transaction active (which is the case in ordinary application
-     *    checkpoints but not all internal cases),
-     * 4. the first dirty update on the page is sufficiently recent the checkpoint transaction would
-     *     skip them,
-     * 5. there's already an address for every disk block involved.
+     *  - This checkpoint happens during RTS, recovery or shutdown. Those scenarios should not leave
+     * anything dirty behind.
+     *  - This is the history store btree. As part of the checkpointing the data store, we will move
+     * the older values into the history store without using any transactions. This led to
+     * representation of all the modifications on the history store page with a transaction that is
+     * maximum than the checkpoint snapshot. But these modifications are done by the checkpoint
+     * itself, so we shouldn't ignore them for consistency.
+     * - If we got to this point and we are dealing with an internal page, this means at least one
+     * of its leaf pages has been reconciled and we need to process the internal page as well.
+     * - There is no snapshot transaction active. Usually, there is one in ordinary application
+     * checkpoints but not all internal cases. Furthermore, this guarantees the metadata file is
+     * never skipped.
+     * - the checkpoint's snapshot includes the first dirty update on the page.
+     * - Not every disk block involved has a disk address.
      */
+    if (F_ISSET(session, WT_SESSION_ROLLBACK_TO_STABLE))
```

Comment:
We may want to shuffle the checks to improve performance as the first few checks are less likely true than the others.

Reply:
Yes, I agree to refactor. We want to return false information as soon as possible.

Reply:
Updated in 0f9a21e1dbf6522f379ab202babc68dc37a09e8a. I don't see considerable improvements and what I see could be noise, I still think it's worth doing. See the jira ticket for the results.